File: singlyLinkedList/singly_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

    def __repr__(self):
        return f'<Node: {self.value}>'


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length += 1
        return self

    def pop(self):
        if not self.head:
            return None
        current = self.head
        current_tail = current
        if self.head == self.tail:
            self.head = None
            self.tail = None
            self.length = 0
        else:
            while current.next:
                current_tail = current
                current = current.next
            self.tail = current_tail
            self.tail.next = None
            self.length -= 1

        return current

    def unshift(self, value):
        node = Node(value)
        if not self.head:
            self.head = node
            self.tail = node
        else:
            current = self.head
            self.head = node
            self.head.next = current
        self.length += 1
        return self

    def shift(self):
        if not self.head:
            return None

        current = self.head
        self.head = current.next

        if not self.head:
            self.tail = None
        self.length -= 1
        return current

    def get_at(self, index):
        if not (0 <= index < self.length):
            return None
        current = self.head
        count = 0

        while count != index:
            current = current.next
            count += 1
        return current

    def set_at(self, index, value):
        node  = self.get_at(index)
        if node:
            node.value = value
            return True
        return False

    def insert(self, index, value):
        if not (0 <= index < self.length):
            return False
        if index == 0:
            return bool(self.unshift(value))
        if index == self.length:
            return bool(self.append(value))

        prev_node = self.get_at(index - 1)
        node = Node(value)
        next_node = prev_node.next
        node.next = next_node
        prev_node.next = node
        self.length += 1
        return True

    def remove(self, index):
        if not (0 <= index < self.length):
            return False

        if index == 0:
            return bool(self.shift())
        if index == (self.length - 1):
            return bool(self.pop())

        prev_node  = self.get_at(index - 1)
        remove_node = prev_node.next
        prev_node.next = remove_node.next
        self.length -= 1
        return True

    def reverse(self):
        node = self.head
        self.head = self.tail
        self.tail = node
        prev = None

        for _ in range(self.length):
            next_node = node.next
            node.next = prev
            prev = node
            node = next_node
        return self


    def __repr__(self):
        items = []
        current = self.head
        while current:
            items.append(str(current.value))
            current = current.next

        return ' -> '.join(items)

File: singlyLinkedList/singly_linked_list.rb
class Node
    attr_accessor :value, :next

    def initialize(value)
        @value = value
        @next = nil
    end

    def to_s()
        "#{@value}"
    end
end

class SinglyLinkedList
    def initialize()
        @head = nil
        @tail = nil
        @size = 0
    end

    def empty?
        if @head == nil
            return true
        end
        return false
    end

    def push(val)
        node  = Node.new(val)

        if self.empty?
            self.init(node)
        else
            @tail.next = node
            @tail = node
        end
        @size += 1
        return self
    end

    def pop
        return nil if self.empty?
        current = @head
        if @head == @tail
            self.clear
            return current
        end
        while current.next
            current_tail = current
            current = current.next
        end
        @tail = current_tail
        @tail.next = nil
        @size -= 1
        return current
    end

    def unshift(value)
        node = Node.new(value)

        if self.empty?
            self.init(node)
        else
            node.next = @head
            @head = node
        end
        @size += 1
        return self
    end

    def shift
        return nil if self.empty?
        current = @head
        if @head == @tail
            self.clear
            return current
        end
        @head = @head.next
        @size -= 1
        return current
    end

    def get(index)
        return nil if not (0 <= index && index < @size)
        return nil if self.empty?

        count = 0
        current = @head
        while count != index
            current = current.next
            count += 1
        end
        return current
    end

    def set(index, value)
        node = self.get(index)
        if node
            node.value = value
            return true
        end
        return false
    end

    def insert(index, value)
        if index == 0
            return self.unshift(value)
        end

        if index == @size
            return self.push(value)
        end

        prev_node = self.get(index - 1)
        if prev_node
            node = Node.new(value)
            tmp = prev_node.next
            node.next = tmp
            prev_node.next = node
            @size += 1
            return self
        end
        return nil
    end

    def remove(index)
        if index == 0
            return self.shift
        end

        if index == @size - 1
            return self.pop
        end

        prev_node = node.get(index - 1)
        if prev_node
            del_node = prev_node.next
            prev_node.next = del_node.next
            @size -= 1
            return del_node
        end
        return  nil
    end

    def reverse
        current = @head
        @head = @tail
        @tail = current
        prev = nil
        while current
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node
        end
        return self
    end

    def to_s
        current = @head
        arr = []
        while current
            arr << current.value
            current = current.next
        end
        arr.join(" -> ")
    end

    def clear
        @head = nil
        @tail = nil
        @size = 0
        return self
    end

    def init(node)
        @head = node
        @tail = node
    end

    protected :clear, :init
end

